server/src/main/java/org/opensearch/indices/recovery/RecoveryState.java

```diff
@@ -56,71 +57,12 @@
  *
  * @opensearch.internal
  */
-public class RecoveryState implements ToXContentFragment, Writeable {
-
-    /**
-     * The stage of the recovery state
-     *
-     * @opensearch.internal
-     */
-    public enum Stage {
-        INIT((byte) 0),
-
-        /**
-         * recovery of lucene files, either reusing local ones are copying new ones
-         */
-        INDEX((byte) 1),
-
-        /**
-         * potentially running check index
-         */
-        VERIFY_INDEX((byte) 2),
-
-        /**
-         * starting up the engine, replaying the translog
-         */
-        TRANSLOG((byte) 3),
-
-        /**
-         * performing final task after all translog ops have been done
-         */
-        FINALIZE((byte) 4),
-
-        DONE((byte) 5);
-
-        private static final Stage[] STAGES = new Stage[Stage.values().length];
-
-        static {
-            for (Stage stage : Stage.values()) {
-                assert stage.id() < STAGES.length && stage.id() >= 0;
-                STAGES[stage.id] = stage;
-            }
-        }
-
-        private final byte id;
-
-        Stage(byte id) {
-            this.id = id;
-        }
-
-        public byte id() {
-            return id;
-        }
-
-        public static Stage fromId(byte id) {
-            if (id < 0 || id >= STAGES.length) {
-                throw new IllegalArgumentException("No mapping for id [" + id + "]");
-            }
-            return STAGES[id];
-        }
-    }
+public class RecoveryState extends ReplicationState implements ToXContentFragment, Writeable {
```

Comment:
Why can't we have two separate Recovery and Replication states without moving Recovery state into Replication State. For instance there are other modes of Recovery like Snapshot recovery that probably might not be termed as replication

Reply:
I agree the inheritance model is a bit confusing.  I know this comes from trying to reduce duplication between segrep and peer recovery paths, but replication will not go through all of these stages.

I'm thinking that we can conceptually remove the inheritance of a "Recovery" target from "ReplicationTarget" in addition to state.  The pieces segrep will reuse here are the lifecycle steps (cancel, done, fail, retry & managing cancellableThreads) and not the methods from `RecoveryTargetHandler`.  This would make it something like "ShardEventTarget"?  I think this would open the door to refactoring RecoveriesCollection into a reusable component as well, which segrep will need a version of.

Reply:
My $0.02 - the inheritance model seems confusing due to how some of these classes are named and the overloaded meaning behind some terms. What we're trying to express with this structure is that there is a common underlying mechanic to these two situations - that of copying segment files. We're choosing to call this common functionality "replication" (since segment files are being replicated) but I can see how that could be confused with "shard replication" between primaries and replicas. 

Further, this file-copying mechanic is reused by peer recovery, but its related classes (like RecoverySource and RecoveryTarget) don't clarify that they are only used for peer recovery.

I don't have any suggestions on how to express this better, so I'm completely open to feedback :)